icicle_v3/backend/cpu/src/curve/cpu_msm.hpp

```diff
@@ -0,0 +1,667 @@
+   * potential collision that requires more urgent ec-addition of the result and the current stored value).
+   * @param task_bkt_idx - address in m_buckets in which to store the result in the future.
+   * @param bkt - the P from m_buckets.
+   * @param base - the P from the input bases
+   * @param negate_base - flag to signal the task to subtract base instead of adding it.
+   */
+  void phase1_push_addition(const unsigned int task_bkt_idx, const P bkt, const A base);
+  /**
+   * @brief Handles the final results of phase 1 (after no other planned additions are required).
+   * The function also handles the potential collision similarly to push_addition above.
+   */
+  void phase1_wait_for_completion();
+
+  /**
+   * @brief Phase 2 of MSM. Function handling the initial parallel part of summing the bucket modules. It splits the
+   * BMs into segments, summing each separately and passing the segments sum to be handled by the final accumulator.
+   * @return vector containing segments line and triangle sums for the final accumulator.
+   */
+  void bm_sum(std::shared_ptr<std::vector<BmSumSegment>>& segments_ptr);
+
+  /**
+   * @brief Setting up phase 2 class members according to phase 1 results.
+   * NOTE: This setup function also potentially blocks until the previous MSM's phase 3 finishes (Given that in batched
+   * MSM, phase 3 is initiated on a separate thread while the next MSM begins phase 1).
+   */
+  void phase2_setup(std::vector<BmSumSegment>& segments);
+
+  /**
+   * @brief Final accumulation required for MSM calculation. the function will either launch a thread to perform the
+   * calculation (`phase3_tread` function) or by the main thread, depending on the position in the batch (Last msm or
+   * not).
+   * @param segments_ptr - Per to vector containing the segment calculations of phase 2
+   * @param idx_in_batch - idx of the current MSM in the batch.
+   * @param result - output, Per to write the MSM result to. Memory for the Per has already been allocated by
+   * the user.
+   */
+  void final_accumulator(std::shared_ptr<std::vector<BmSumSegment>>& segments_ptr, int idx_in_batch, P* result);
+
+  /**
+   * @brief Phase 3 function to be ran by a separate thread (or main in the end of the run) - i.e. without using the
+   * tasks manager. It performs the remaining serial addition in each BM and sums them to one final MSM result.
+   * @param result - Per to write the MSM result to. Memory for the Per has already been allocated by the user.
+   */
+  void phase3_thread(std::shared_ptr<std::vector<BmSumSegment>> segments_ptr, P* result);
+
+  /**
+   * @brief Function for resetting class members between batch runs.
+   */
+  void batch_run_reset() { std::fill(m_bkts_occupancy.begin(), m_bkts_occupancy.end(), false); }
+};
+
+template <typename A, typename P>
+Msm<A, P>::Msm(const MSMConfig& config)
+    : manager(
+        config.ext->get<int>("n_threads") > 0 ? config.ext->get<int>("n_threads")
```

Comment:
(1) note that ext may be null. You must check it before call get.
(2) add the n_threads to https://github.com/ingonyama-zk/icicle/blob/yshekel/V3/icicle_v3/include/icicle/backend/msm_config.h 
(3) use the value from (2) rather than hard coded string.

Reply:
1+2 handled